stratum/hal/lib/barefoot/bf_sde_interface.h

```diff
@@ -71,6 +91,86 @@ class BfSdeInterface {
   // Check whether we are running on the software model.
   virtual ::util::StatusOr<bool> IsSoftwareModel(int device) = 0;
 
+  virtual ::util::Status TxPacket(int device, const std::string& packet) = 0;
+
+  virtual ::util::Status StartPacketIo(int device) = 0;
+
+  virtual ::util::Status StopPacketIo(int device) = 0;
+
+  virtual ::util::Status RegisterPacketReceiveWriter(
+      int device, std::unique_ptr<ChannelWriter<std::string>> writer) = 0;
+
+  virtual ::util::Status UnregisterPacketReceiveWriter(int device) = 0;
+
+  // Create a new multicast node with the given parameters. Returns the newly
+  // allocated node id.
+  virtual ::util::StatusOr<uint32> CreateMulticastNode(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      int mc_replication_id, const std::vector<uint32>& mc_lag_ids,
+      const std::vector<uint32> ports) = 0;
+
+  // Returns the node IDs linked to the given multicast group ID.
+  // TODO(max): rename to GetMulticastNodeIdsInMulticastGroup
+  virtual ::util::StatusOr<std::vector<uint32>> GetNodesInMulticastGroup(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      uint32 group_id) = 0;
+
+  // Delete the given multicast nodes.
+  virtual ::util::Status DeleteMulticastNodes(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      const std::vector<uint32>& mc_node_ids) = 0;
+
+  // Returns the multicast node with the given ID ($pre.node table).
+  virtual ::util::Status GetMulticastNode(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      uint32 mc_node_id, int* replication_id,
+      std::vector<uint32>* lag_ids, std::vector<uint32>* ports) = 0;
+
+  // Inserts a multicast group ($pre.mgid table).
+  virtual ::util::Status InsertMulticastGroup(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      uint32 group_id, const std::vector<uint32>& mc_node_ids) = 0;
+
+  // Modifies a multicast group ($pre.mgid table).
+  virtual ::util::Status ModifyMulticastGroup(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      uint32 group_id, const std::vector<uint32>& mc_node_ids) = 0;
+
+  // Deletes a multicast group ($pre.mgid table).
+  virtual ::util::Status DeleteMulticastGroup(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      uint32 group_id) = 0;
+
+  // Returns the multicast group with the given ID ($pre.mgid table), or all
+  // groups if ID is 0.
+  virtual ::util::Status GetMulticastGroups(
+      int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+      uint32 group_id, std::vector<uint32>* group_ids,
+      std::vector<std::vector<uint32>>* mc_node_ids) = 0;
```

Comment:
This function interface is not the greatest, but least bad alternative I could find.